### PR TITLE
Added more error handling into the API, as well as retry to flushing events queue.

### PR DESCRIPTION
**Pull request by kiddinn**
This PR adds more error handling to survive a search index that is not found in Elastic, as well as adding a retry count to the flushing of the event queue in case there is a socket timeout.

### CODE REVIEW

File: timesketch/api/v1/resources/upload.py [hunk 1]
```diff
@@ -133,6 +133,14 @@ def _upload_and_index(
...
             data_label=data_label,
             extension=file_extension)
+
+        if not searchindex:
+            abort(
+                HTTP_STATUS_CODE_BAD_REQUEST,
+                'We were unable to acquire a searchindex, and therefore not '
+                'able to upload data, try again? Please create an issue on '
+                'Github.')
+
```
Comment (berggren) [inline timesketch/api/v1/resources/upload.py:143]:
```suggestion
                HTTP_STATUS_CODE_BAD_REQUEST,
                'We were unable to acquire a searchindex and therefore not '
                'able to upload data, please try again. If this error persist please create an issue on '
                'Github: https://github.com/google/timesketch/issues/new/choose')

```

File: timesketch/lib/datastores/elastic.py [hunk 2]
```diff
@@ -912,14 +926,15 @@ def import_event(self, index_name, event_type, event=None, event_id=None,
 
         return self.import_counter['events']
 
-    def flush_queued_events(self):
+    def flush_queued_events(self, retry_count=0):
```
Comment (berggren) [inline timesketch/lib/datastores/elastic.py:929]: Why have retry_count as an argument? Does it get set from a caller somewhere?

File: timesketch/lib/utils.py [hunk 3]
```diff
@@ -83,6 +84,31 @@ def _scrub_special_tags(dict_obj):
...
+    Args:
+        indices (list): List of indices.
+        datastore (ElasticsearchDataStore): a data store object.
+
+    Returns:
+        list of indices that are discovered.
+    """
+    for index in indices:
+        try:
+            datastore.client.indices.refresh(index=index)
```
Comment (berggren) [inline timesketch/lib/utils.py:102]: How expensive is this? We are doing this on every search. Is it maybe enough to do this check in the Sketch (GET) resource?

**Review by berggren — CHANGES_REQUESTED** [review]

File: timesketch/lib/datastores/elastic.py (hunk 2 above)
Comment (kiddinn) [reply timesketch/lib/datastores/elastic.py:929]: it is called by the exception handling, only there, to be a `recursive` function

File: timesketch/lib/utils.py (hunk 3 above)
Comment (kiddinn) [reply timesketch/lib/utils.py]: I measured it, it seems fairly fast, but I also checked the `if index exists` and it seems about the same speed, but then a lot less overhead with the try/except, so I swapped the implementation.

**Review by kiddinn — COMMENTED** [review]
PTAL

File: timesketch/lib/utils.py [hunk 4]
```diff
@@ -83,6 +84,23 @@ def _scrub_special_tags(dict_obj):
             _ = dict_obj.pop(field)
 
 
+def validate_list_of_indices(indices, datastore):
```
Comment (berggren) [inline timesketch/lib/utils.py:87]:
```suggestion
def validate_indices(indices, datastore):
```

File: timesketch/api/v1/resources/explore.py [hunk 5]
```diff
@@ -103,10 +104,18 @@ def post(self, sketch_id):
         if '_all' in indices:
             indices = all_indices
 
+        # Remove indices that don't exist from search.
+        indices = utils.validate_list_of_indices(indices, self.datastore)
```
Comment (berggren) [inline timesketch/api/v1/resources/explore.py:108]:
```suggestion
        indices = utils.validate_indices(indices, self.datastore)
```

**Review by berggren — CHANGES_REQUESTED** [review]

**Review by berggren — APPROVED** [review]

**Pull request merged**